### ISSUE TITLE
stackalloc is causing bounds check elimination to break

### ISSUE THREAD

**Issue by stephentoub** [dotnet/runtime#55777]
[sharplab.io](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIAYACY8gOgCUBXAOwwEt8YLAMIR8AB14AbGFADKMgG68wMXAG4a9Jq049+ggJI8ZEMfKhKV6mjQDa+CABMO0kA1kBrXmIAyEMNiSuEa8GAC6NrQAzEykDEI0AN40DKnaSAzAEBCSDACy5AAUbDDYjgDyXJIAnrJi2FwAPGAAFthQAHwMvFxiHBgAlClpydRp490AZgyFPX0YLD4wXADmGC0MAIQAvAykAwzDExPEAOwMk4G4MBrURxN1Dc1tnQywuC4YDLu4GNhgHkCkn8DFa7VspDCt2OqXen1sdDC326vX6CKh93GcMkGFs5CRuzmaPx0JhZwYGCgHBu9wAvvd7swMlkcvlSMVShUqrV6k0wa8iYN7qMYVMZoLFss1hsdnsDpiyedLkEaXcxjChC84vzSKTjgBVLi4bCTQSebwhDCFCD9UFagZ6h6857tLrYr67PIwBxQap5dq4NqSYSwbAYGCPLiFWDTHUsWR/KAYNByx1Y1TwxHIwXotNpd14gko+aFvOpcmU6mO+nUGt2BNUsAYHzYao2q0ttv9ADSPUc8ZgAEdqXpAinZLwAF4wZEoAYRai/RtfTXtOKJJgxfnuRMYNQMGtAA===)

```C#
using System;
using System.Runtime.CompilerServices;
using System.Runtime.InteropServices;

[module: SkipLocalsInit]

public class C
{
    static bool M1(ReadOnlySpan<char> input)
    {
        if (input.Length != 2) 
            return false;

        Span<char> result = stackalloc char[2];
        result[0] = input[0];
        result[1] = input[1];
        return true;
    }
    
    static bool M2(ReadOnlySpan<char> input)
    {
        if (input.Length != 2) 
            return false;

        Char2 char2;
        Unsafe.SkipInit(out char2);
        Span<char> result = MemoryMarshal.CreateSpan(ref char2.Start, 2);
        result[0] = input[0];
        result[1] = input[1];
        return true;
    }
}

[StructLayout(LayoutKind.Sequential, Size = 4)]
struct Char2 { public char Start; }

```
results in:
```
; Core CLR v5.0.721.25508 on amd64

C..ctor()
    L0000: ret

C.M1(System.ReadOnlySpan`1<Char>)
    L0000: sub rsp, 0x38
    L0004: mov rax, 0xd0f358f224d5
    L000e: mov [rsp+0x30], rax
    L0013: mov edx, [rcx+8]
    L0016: cmp edx, 2
    L0019: je short L0039
    L001b: xor eax, eax
    L001d: mov rcx, 0xd0f358f224d5
    L0027: cmp [rsp+0x30], rcx
    L002c: je short L0033
    L002e: call 0x00007ffc6346bda0
    L0033: nop
    L0034: add rsp, 0x38
    L0038: ret
    L0039: lea rax, [rsp+0x28]
    L003e: cmp edx, 0
    L0041: jbe short L007e
    L0043: mov rdx, [rcx]
    L0046: movzx edx, word ptr [rdx]
    L0049: mov [rax], dx
    L004c: cmp dword ptr [rcx+8], 1
    L0050: jbe short L007e
    L0052: mov rdx, [rcx]
    L0055: movzx edx, word ptr [rdx+2]
    L0059: mov [rax+2], dx
    L005d: mov eax, 1
    L0062: mov rcx, 0xd0f358f224d5
    L006c: cmp [rsp+0x30], rcx
    L0071: je short L0078
    L0073: call 0x00007ffc6346bda0
    L0078: nop
    L0079: add rsp, 0x38
    L007d: ret
    L007e: call 0x00007ffc6346a5e0
    L0083: int3

C.M2(System.ReadOnlySpan`1<Char>)
    L0000: push rax
    L0001: mov rdx, [rcx]
    L0004: mov eax, [rcx+8]
    L0007: cmp eax, 2
    L000a: je short L0013
    L000c: xor eax, eax
    L000e: add rsp, 8
    L0012: ret
    L0013: lea rax, [rsp]
    L0017: movzx ecx, word ptr [rdx]
    L001a: mov [rax], cx
    L001d: movzx edx, word ptr [rdx+2]
    L0021: mov [rax+2], dx
    L0025: mov eax, 1
    L002a: add rsp, 8
    L002e: ret
```

cc: @EgorBo 

category:correctness
theme:stack-allocation

**Comment by MichalPetryka** [comment]
M2 will cause a stack corruption due to Char2 being 2 bytes and the span created on it 4 bytes, is this intended? I assume the Size in StructLayout was supposed to be 4, not 2.

**Comment by stephentoub** [comment]
That was just a typo and doesn't change the core issue.

**Comment by SingleAccretion** [comment]
Noting the .NET 6 codegen has only one check:
```asm
; Method RyuJitReproduction.Program:BoundsChecksRaw(System.ReadOnlySpan`1[Char]):bool
G_M9484_IG01:
       sub      rsp, 56
       xor      eax, eax
       mov      qword ptr [rsp+28H], rax
       mov      rax, 0xD1FFAB1E
       mov      qword ptr [rsp+30H], rax
G_M9484_IG02:
       mov      eax, dword ptr [rcx+8]
       cmp      eax, 2
       je       SHORT G_M9484_IG06
G_M9484_IG03:
       xor      eax, eax
       mov      rcx, 0xD1FFAB1E
       cmp      qword ptr [rsp+30H], rcx
       je       SHORT G_M9484_IG04
       call     CORINFO_HELP_FAIL_FAST
G_M9484_IG04:
       nop      
G_M9484_IG05:
       add      rsp, 56
       ret      
G_M9484_IG06:
       lea      rax, bword ptr [rsp+28H]
       mov      rdx, bword ptr [rcx]
       movzx    rdx, word  ptr [rdx]
       mov      word  ptr [rax], dx
       cmp      dword ptr [rcx+8], 1
       jbe      SHORT G_M9484_IG09
       mov      rdx, bword ptr [rcx]
       movzx    rdx, word  ptr [rdx+2]
       mov      word  ptr [rax+2], dx
       mov      eax, 1
       mov      rcx, 0xD1FFAB1E
       cmp      qword ptr [rsp+30H], rcx
       je       SHORT G_M9484_IG07
       call     CORINFO_HELP_FAIL_FAST
G_M9484_IG07:
       nop      
G_M9484_IG08:
       add      rsp, 56
       ret      
G_M9484_IG09:
       call     CORINFO_HELP_RNGCHKFAIL
       int3     
; Total bytes of code: 134
```

Fundamentally the issue is caused by the fact that we give up on promoting `input`:
```scala
  struct promotion of V00 is disabled because lvIsParam and compGSReorderStackLayout
```

Which then leads to optimizer seeing trees like this for the `_length` field:
```scala
***** BB02
STMT00001 (IL   ???...  ???)
N004 (  4,  4) CSE #01 (def)[000087] *---G-------                 +--*  IND       int    <l:$200, c:$240>
N003 (  2,  2)              [000129] -------N----                 |  \--*  ADD       byref  $1c0
N001 (  1,  1)              [000086] -------N----                 |     +--*  LCL_VAR   byref  V15 tmp12        u:2 $80
N002 (  1,  1)              [000128] ------------                 |     \--*  CNS_INT   long   8 field offset Fseq[_length] $180
```

However, why do we see only one check?

Well that's the optimizer being smart :). The sequence of events goes something like this:
1. CSE promotes _two_ uses of `input.Length` to a variable which the optimizer can reason about. Why only two instead of all three? It is because of memory SSA & VN:
```scala
ByrefExposedMemory VN = $100
if (*input.Length ...) <- use+def #1

// Check №1
if (*input.length ...) <- use #2

// We are mutating ByrefExposedMemory!
*result[0] = input[0];
ByrefExposedMemory VN = $440

// Check №2 now sees different VN for *input.Length, since it is a ByrefExposedLoad,
// and the length is not CSEd, leading to no check elimination as well.
```
2. Assertion propagation and range check see that the CSEd variable is compared against the constant bound and eliminate the first check.

Based on this, a couple source-level workarounds can be devised:
1\) Copy the input span:
```cs
static bool BoundsChecksCopy(ReadOnlySpan<char> input)
{
    ReadOnlySpan<char> inputCopy = input;

    if (inputCopy.Length != 2)
        return false;

    Span<char> result = stackalloc char[2];
    result[0] = inputCopy[0];
    result[1] = inputCopy[1];
    return true;
}
```
```asm
G_M21869_IG01:
       sub      rsp, 24
       xor      eax, eax
       mov      qword ptr [rsp+08H], rax
       mov      rax, 0xD1FFAB1E
       mov      qword ptr [rsp+10H], rax
G_M21869_IG02:
       mov      rdx, bword ptr [rcx]
       mov      eax, dword ptr [rcx+8]
       cmp      eax, 2
       je       SHORT G_M21869_IG06
G_M21869_IG03:
       xor      eax, eax
       mov      rcx, 0xD1FFAB1E
       cmp      qword ptr [rsp+10H], rcx
       je       SHORT G_M21869_IG04
       call     CORINFO_HELP_FAIL_FAST
G_M21869_IG04:
       nop      
G_M21869_IG05:
       add      rsp, 24
       ret      
G_M21869_IG06:
       lea      rax, bword ptr [rsp+08H]
       movzx    rcx, word  ptr [rdx]
       mov      word  ptr [rax], cx
       movzx    rdx, word  ptr [rdx+2]
       mov      word  ptr [rax+2], dx
       mov      eax, 1
       mov      rcx, 0xD1FFAB1E
       cmp      qword ptr [rsp+10H], rcx
       je       SHORT G_M21869_IG07
       call     CORINFO_HELP_FAIL_FAST
G_M21869_IG07:
       nop      
G_M21869_IG08:
       add      rsp, 24
       ret      
```
2\) Reorder the stores...
```cs
static bool BoundsChecksReorder(ReadOnlySpan<char> input)
{
    if (input.Length != 2)
        return false;

    Span<char> result = stackalloc char[2];
    (result[0], result[1]) = (input[0], input[1]);
    return true;
}
```
```asm
       sub      rsp, 24
       xor      eax, eax
       mov      qword ptr [rsp+08H], rax
       mov      rax, 0xD1FFAB1E
       mov      qword ptr [rsp+10H], rax
G_M22641_IG02:
       mov      eax, dword ptr [rcx+8]
       cmp      eax, 2
       je       SHORT G_M22641_IG06
G_M22641_IG03:
       xor      eax, eax
       mov      rcx, 0xD1FFAB1E
       cmp      qword ptr [rsp+10H], rcx
       je       SHORT G_M22641_IG04
       call     CORINFO_HELP_FAIL_FAST
G_M22641_IG04:
       nop      
G_M22641_IG05:
       add      rsp, 24
       ret      
G_M22641_IG06:
       lea      rax, bword ptr [rsp+08H]
       lea      rdx, bword ptr [rax+2]
       mov      rcx, bword ptr [rcx]
       movzx    r8, word  ptr [rcx]
       movzx    rcx, word  ptr [rcx+2]
       mov      word  ptr [rax], r8w
       mov      word  ptr [rdx], cx
       mov      eax, 1
       mov      rcx, 0xD1FFAB1E
       cmp      qword ptr [rsp+10H], rcx
       je       SHORT G_M22641_IG07
       call     CORINFO_HELP_FAIL_FAST
G_M22641_IG07:
       nop      
G_M22641_IG08:
       add      rsp, 24
       ret  
```

How to fix this in the compiler requires more careful thinking, obviously.

**Comment by jakobbotsch** [comment]
This was fixed by physical promotion (#88090).
